### PR TITLE
Basic genreRecommendation functions

### PR DESCRIPTION
**Pull request by ankitamitra**
Based on [algorithm design](https://docs.google.com/document/d/1DjPFQNI3uuV751hqZb_bMrVB4Yi-xhCF4q7ch83ULgI/edit?usp=sharing):

- Constructor for recommender class
- Get genreSet given a Book
- Get bookSet given set of Genres (has to be an exact match)

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/data/Book.java [hunk 1]
```diff
@@ -27,16 +28,17 @@ public Builder addReview(String review) {
...
+     * The two replacements in the first line are in fact, different
+     * If anyone knows why or how to fix this in a cleaner way, please let ankita know!
      **/
     private String cleanReview(String original) {
-        String newString = original.replaceAll("â€", "'").replaceAll("â€", "'");
-        newString = newString.replaceAll("[^a-zA-Z0-9, .:*?!'#<>(){}/-]", "");
-        return newString;
+      String newString = original.replaceAll("â€", "'").replaceAll("â€", "'");
+      newString = newString.replaceAll("[^a-zA-Z0-9, .:*?!'#<>(){}/-]", "");
+      return newString;
```
Comment (ankitamitra) [inline portfolio/src/main/java/com/google/sps/data/Book.java:41]: clang-format (no changes from past PR)

File: portfolio/src/main/java/com/google/sps/data/BookFieldsEnum.java [hunk 2]
```diff
@@ -16,15 +16,14 @@
...
+
   private BookFieldsEnum(String javaScriptProperty) {
     this.javaScriptProperty = javaScriptProperty;
   }
- 
+
   public String getJSProperty() {
     return this.javaScriptProperty;
   }
 }
```
Comment (ankitamitra) [inline portfolio/src/main/java/com/google/sps/data/BookFieldsEnum.java:29]: clang-format (no changes from past PR)

File: portfolio/src/main/java/com/google/sps/data/Book.java [hunk 3]
```diff
@@ -27,16 +28,17 @@ public Builder addReview(String review) {
       return this;
     }
 
-    /** In this function, we use replaceAll instead of enconding/decoding
-      * methods because the bad characters do not come in during our encoding, 
-      * but rather, already exist in the file that we're reading from. 
-      * The two replacements in the first line are in fact, different
-      * If anyone knows why or how to fix this in a cleaner way, please let ankita know!
+    /**
+     *In this function, we use replaceAll instead of enconding/decoding
```
Comment (danyag0820) [inline portfolio/src/main/java/com/google/sps/data/Book.java:32]: nit: typo in encoding?

File: portfolio/src/main/java/com/google/sps/data/GenreRecommender.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  /**
+   * Constructor that takes in a list of books
+   * and populates class variables appropriately.
+   *
+   * @param books: List of Book objects for the recommendation algorithm
+   */
+  public GenreRecommender(List<Book> books) {
+    this.books = books;
+    bookToGenres = new HashMap<Book, Set<String>>();
+    genreToBooks = new HashMap<String, Set<Book>>();
```
Comment (saur-g) [inline portfolio/src/main/java/com/google/sps/data/GenreRecommender.java:29]: nit: (mostly fyi, change is optional) These can be replaced with `ImmutableSetMultiMap1`. It doesn't require you to handle the case whether key already exists in the map or not. It also returns ImmutableSet on get, so you don't have to copy it.

https://guava.dev/releases/19.0/api/docs/com/google/common/collect/ImmutableSetMultimap.html

File: portfolio/src/test/java/com/google/sps/GenreRecommenderTest.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+          .title("book about me")
+          .genre(new HashSet<>(Arrays.asList("autobiography")))
+          .addReview("N/A")
+          .build();
+  private static final Book BIOGRAPHY_BOOK = Book.builder()
+                                                 .title("book about him")
+                                                 .genre(new HashSet<>(Arrays.asList("biography")))
+                                                 .addReview("N/A")
+                                                 .build();
+  private static final Book CRIME_BOOK = Book.builder()
```
Comment (saur-g) [inline portfolio/src/test/java/com/google/sps/GenreRecommenderTest.java:42]: opt: can do in separate PR.

It might be helpful to create a static helper function in this test class to create books instance (to avoid redundant code, and improve readability):
````
private static final Book BIOGRAPHY_BOOK = createBook("book about him", "biography");

private static Book createBook(String title, String ...genres) {
  Book.builder()
    .title(title)
    .genre(new HashSet<>(Arrays.asList(genres)))
    .addReview("N/A")
    .build();
}
```

By convention we define the private methods in class after the test cases.

File: portfolio/src/test/java/com/google/sps/GenreRecommenderTest.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+import org.junit.runners.JUnit4;
+
+/**
+ * This is a test class for BookReader.java
+ * It only tests the basic functionality for reading from a CSV file.
+ * It does not check that the results match those from the Books API.
+ */
+@RunWith(JUnit4.class)
+public final class GenreRecommenderTest {
+  private List<Book> bookList;
```
Comment (saur-g) [inline portfolio/src/test/java/com/google/sps/GenreRecommenderTest.java:28]: nit: May be move this declaration within the test, as it is not user elsewhere. Global variables in test are more useful when they are used in setup/teardown before/after code, or are final/static variables.
